be1-go/hub/organizer.go

```diff
@@ -484,3 +486,203 @@ func (c *laoChannel) processRollCallObject(data message.Data) error {
+	if !ok {
+		return &message.Error{
+			Code:        -4,
+			Description: "failed to cast data to SetupElectionData",
+		}
+	}
+
+	// Check if the Lao ID of the message corresponds to the channel ID
+	encodedLaoID := base64.StdEncoding.EncodeToString(data.LaoID)
+	channelID := c.channelID[6:]
+	if channelID != encodedLaoID {
+		return &message.Error{
+			Code:        -4,
+			Description: fmt.Sprintf("Lao ID of the message (Lao: %s) is different from the channelID (channel: %s)", encodedLaoID, channelID),
+		}
+	}
+
+	// Compute the new election channel id
+	encodedElectionID := base64.StdEncoding.EncodeToString(data.ID)
+	encodedID := encodedLaoID + "/" + encodedElectionID
+
+	// Create the new election channel
+	electionCh := electionChannel{
+		createBaseChannel(o, "/root/"+encodedID),
+		data.StartTime,
+		data.EndTime,
+		false,
+		getAllQuestionsForElectionChannel(data.Questions, data),
+	}
+
+	// Add the SetupElection message to the new election channel
+	messageID := base64.StdEncoding.EncodeToString(msg.MessageID)
+	electionCh.inbox[messageID] = msg
+
+	// Add the new election channel to the organizerHub
+	o.channelByID[encodedID] = &electionCh
+
+	return nil
+}
+
+func getAllQuestionsForElectionChannel(questions []message.Question, data *message.ElectionSetupData) map[string]question {
+	qs := make(map[string]question)
+	for _, q := range questions {
+		qs[base64.StdEncoding.EncodeToString(q.ID)] = question{
+			q.ID,
+			q.BallotOptions,
+			sync.RWMutex{},
+			make(map[string]validVote),
+			q.VotingMethod,
+		}
+	}
+	return qs
+}
+
+type electionChannel struct {
+	*baseChannel
+
+	// Starting time of the election
+	start message.Timestamp
+
+	// Ending time of the election
+	end message.Timestamp
+
+	// True if the election is over and false otherwise
+	terminated bool
+
+	// Questions asked to the participants
+	//the key will be the string representation of the id of type byte[]
+	questions map[string]question
+}
+
+type question struct {
+	// ID of th question
+	id []byte
+
+	// Different options
+	ballotOptions []message.BallotOption
+
+	//valid vote mutex
+	validVotesMu sync.RWMutex
+
+	// list of all valid votes
+	validVotes map[string]validVote
+
+	// Voting method of the election
+	method message.VotingMethod
+}
+
+type validVote struct {
+	// time of the creation of the vote
+	voteTime message.Timestamp
+
+	// indexes of the ballot options
+	indexes []int
+}
+
+func (c *electionChannel) Publish(publish message.Publish) error {
+	err := c.baseChannel.VerifyPublishMessage(publish)
+	if err != nil {
+		return xerrors.Errorf("failed to verify publish message on an election channel: %v", err)
+	}
+
+	msg := publish.Params.Message
+
+	data := msg.Data
+
+	object := data.GetObject()
+
+	if object == message.ElectionObject {
+
+		action := message.ElectionAction(data.GetAction())
+		switch action {
+		case message.CastVoteAction:
+			return castVoteHelper(publish, c)
+		case message.ElectionEndAction:
```

Comment:
You might want to log here that it's not implemented yet

Reply:
Agreed